File: lib/ts/src/internal/validation/ValidateResult.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TUnion } from '../types/TUnion.js';
import { TelepactError } from '../../TelepactError.js';
import { mapValidationFailuresToInvalidFieldCases } from '../../internal/validation/MapValidationFailuresToInvalidFieldCases.js';
import { ValidateContext } from './ValidateContext.js';

export function validateResult(resultUnionType: TUnion, errorResult: any): void {
    const newErrorResultValidationFailures = resultUnionType.validate(errorResult, [], new ValidateContext(null, null, false));
    if (newErrorResultValidationFailures.length !== 0) {
        throw new TelepactError(
            `Failed internal telepact validation: ${JSON.stringify(mapValidationFailuresToInvalidFieldCases(newErrorResultValidationFailures))}`,
        );
    }
}

File: lib/go/internal/validation/ValidateResult.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package validation

import (
	"fmt"

	telepact "github.com/telepact/telepact/lib/go"
	"github.com/telepact/telepact/lib/go/internal/types"
)

// ValidateResult ensures the provided error result conforms to the union type definition.
func ValidateResult(resultUnionType *types.TUnion, errorResult map[string]any) error {
	if resultUnionType == nil {
		return telepact.NewTelepactError("result union type is nil")
	}

	validateCtx := types.NewValidateContext(nil, "", false)
	validationFailures := resultUnionType.Validate(errorResult, nil, validateCtx)
	if len(validationFailures) == 0 {
		return nil
	}

	cases := types.MapValidationFailuresToInvalidFieldCases(validationFailures)
	return telepact.NewTelepactError(fmt.Sprintf("Failed internal telepact validation: %v", cases))
}
